File: 20. Maximum of all subarrays of size k.py
"""
Task: Maximum of all subarrays of size k (sliding window maximum)
Time Complexity: O(n)
Space Complexity: O(k)
"""

from collections import deque


def max_subarrays(a, k):
    dq = deque()
    res = []
    for i, v in enumerate(a):
        while dq and dq[0] <= i - k:
            dq.popleft()
        while dq and a[dq[-1]] <= v:
            dq.pop()
        dq.append(i)
        if i >= k - 1:
            res.append(a[dq[0]])
    return res

if __name__ == "__main__":
    print(max_subarrays([1,3,-1,-3,5,3,6,7], 3))

# Task: Maximum of all subarrays of size k

File: 20. Maximum of all subarrays of size k.cpp
/*
 Task: Maximum of all subarrays of size k (same as sliding window maximum)
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> maxSubarrays(vector<int>& a, int k)
{
    deque<int> dq;
    vector<int> res;

    for (int i = 0; i < static_cast<int>(a.size()); ++i)
    {
        while (!dq.empty() && dq.front() <= i - k) dq.pop_front();

        while (!dq.empty() && a[dq.back()] <= a[i]) dq.pop_back();

        dq.push_back(i);

        if (i >= k - 1)
            res.push_back(a[dq.front()]);
    }

    return res;
}

int main()
{
    vector<int> a = {1, 3, -1, -3, 5, 3, 6, 7};
    for (auto x : maxSubarrays(a, 3)) cout << x << " ";
    cout << "\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(k)
